### PR TITLE
refactor(sunjx): refactor loss-filter implementation

### PR DESCRIPTION
**Pull request by Jiaxuan-Sun**
Add new `lightrft/trainer/filter_weight/` module with:

- **`metrics.py`** - Metrics computation layer (entropy, difficulty, staleness, etc.)
- **`filters.py`** - Sample filtering layer (length, reward value, entropy, difficulty filters, etc.)
- **`weights.py`** - Loss weighting layer (length, entropy, difficulty, staleness weightings, etc.)
- **`manager.py`** - Unified management layer (FilterWeightManager)

Note:​ The dynamic sampling feature has been tested. Other components are reserved for future extension.

### CODE REVIEW

File: lightrft/trainer/filter_weight/__init__.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+    """Get the version of the filter_weight module."""
+    return __version__
+
+
+# Quick access functions for common use cases
+def create_length_filter(max_length: int = 1024, **kwargs):
+    """
+    Quick function to create a response length filter.
+
+    Args:
```
Comment (puyuan1996) [inline lightrft/trainer/filter_weight/__init__.py]: Please use the :param style docstring format.

File: lightrft/trainer/filter_weight/filters.py [hunk 2]
```diff
@@ -0,0 +1,426 @@
+"""
+Sample Filtering Module
+
+Provides unified interface for filtering samples based on various criteria.
+
+Author: LightRLHF Team
+"""
```
Comment (puyuan1996) [inline lightrft/trainer/filter_weight/filters.py]: Please remove the Author line.

File: lightrft/trainer/filter_weight/__init__.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+
+    # ========== Manager ==========
+    "FilterWeightManager",
+    "FilterWeightManagerBuilder",
+]
+
+
+# Version info
+__version__ = "1.0.0"
+__author__ = "LightRLHF Team"
```
Comment (puyuan1996) [inline lightrft/trainer/filter_weight/__init__.py]: Remove the wrong version info

**Review by puyuan1996 — CHANGES_REQUESTED** [review]

File: lightrft/trainer/filter_weight/__init__.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+    )
+
+    # Compute metrics
+    metrics = manager.compute_metrics(outputs)
+
+    # Apply to experiences
+    experiences, weights = manager.apply_to_experiences(experiences, metrics)
+    ```
+
+Author: LightRLHF Team
```
Comment (puyuan1996) [inline lightrft/trainer/filter_weight/__init__.py]: Please remove the Author line.

File: lightrft/trainer/filter_weight/manager.py [hunk 5]
```diff
@@ -0,0 +1,463 @@
...
+            EntropyWeighting,
+            DifficultyWeighting,
+            StalenessWeighting,
+        )
+
+        # Build filters
+        filters = []
+
+        # Response length filter (from overlong_buffer settings)
+        if getattr(args, "overlong_buffer", False):
```
Comment (puyuan1996) [inline lightrft/trainer/filter_weight/manager.py]: Avoid using getattr. Instead, please add the corresponding default parameters to the config definition. This allows us to access them directly via args.xxx and prevents issues where parameters are silently assigned incorrect default values.

File: lightrft/trainer/filter_weight/metrics.py [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+            # In practice, you'd need to add generation_step to _SamplesOutput
+            pass
+
+        return SampleMetrics(
+            response_length=response_lengths,
+            entropy=entropy,
+            logit_kl=logit_kl,
+            difficulty=difficulty,
+            staleness=staleness,
+            reward_value=rewards,
```
Comment (puyuan1996) [inline lightrft/trainer/filter_weight/metrics.py]: Have we tested the filters related to logit_kl, difficulty, and staleness? If enabled, how exactly should they be used? We need to clarify which metrics have been tested, which ones still need testing, and which are theoretically pending/undetermined.

File: lightrft/strategy/strategy_base.py [hunk 7]
```diff
@@ -416,6 +416,17 @@ def all_reduce(self,
...
+            # Ensure consistent key order and presence across ranks to avoid NCCL hangs.
+            if dist.is_available() and dist.is_initialized() and self.world_size > 1:
+                local_keys = list(data.keys())
+                gathered_keys = [None for _ in range(self.world_size)]
+                dist.all_gather_object(gathered_keys, local_keys)
+                all_keys = sorted({k for keys in gathered_keys for k in keys})
+                ret = {}
+                for k in all_keys:
+                    ret[k] = self.all_reduce(data.get(k, 0.0), op)
+                return ret
```
Comment (puyuan1996) [inline lightrft/strategy/strategy_base.py]: Why was this added? Does it cause an error without it?

File: lightrft/trainer/fast_exp_maker.py [hunk 8]
```diff
@@ -1407,7 +1441,15 @@ def _process_experiences(
...
         # ========== Overlong Sequence Penalty ==========
-        if config.overlong_buffer:
+        # Use new filter_weight framework if enabled, otherwise use legacy logic
+        from .filter_weight import ResponseLengthFilter
+        use_filter_weight = (
+            self.filter_weight_manager is not None and self.filter_weight_manager.filters
+            and any(isinstance(f, ResponseLengthFilter) for f in self.filter_weight_manager.filters)
+        )
+
+        if config.overlong_buffer and not use_filter_weight:
```
Comment (puyuan1996) [inline lightrft/trainer/fast_exp_maker.py]: Shouldn't the overlong buffer penalty be merged into the length filter/weight logic?

**Review by puyuan1996 — CHANGES_REQUESTED** [review]

File: lightrft/strategy/strategy_base.py (hunk 7 above)
Comment (Jiaxuan-Sun) [reply lightrft/strategy/strategy_base.py]: This is to prevent deadlock in distributed all-reduce operations.  
After dynamic sampling, the set of keys in the status dictionary may differ across ranks (some ranks have keys like kl and ptx_loss, while others do not). The all_reduce(dict) operation calls dist.all_reduce for each key individually. If the keys or their order differ between ranks, the collective operations will be inconsistent, causing the process to hang.

File: lightrft/trainer/ppo_trainer_vl.py [hunk 9]
```diff
@@ -737,24 +742,34 @@ def training_step_actor(self,
...
-                kl = compute_approx_kl(
-                    action_log_probs,
-                    base_action_log_probs,
-                    experience.action_mask,
-                    kl_estimator=self.args.kl_estimator,
-                )
+                # If no valid actions or base log-probs are empty, skip KL safely.
+                if ((experience.action_mask is not None and experience.action_mask.sum().item() == 0)
+                        or (base_action_log_probs is not None and base_action_log_probs.numel() == 0)):
+                    kl = torch.zeros_like(
```
[the remark this answers is not in the document]
Comment (puyuan1996) [reply lightrft/trainer/ppo_trainer_vl.py:763]: If that's the case, we should probably figure out a way to avoid such issues during the upstream filter/weight stages (e.g., by filtering out these invalid batches early on), rather than just forcing the KL to 0 here. Setting it to 0 is more of a workaround and might mask underlying issues with the data flow or sampling logic.

如果是这样的话，那我们最好想办法在前置的 filter 或 weight 阶段就规避掉这类问题（比如提前把这些无效数据的 batch 过滤掉），而不是在这里强行让 KL 为 0。因为在这里直接置为 0 治标不治本，反而可能会掩盖潜在的数据流转或采样逻辑问题。

File: lightrft/strategy/strategy_base.py (hunk 7 above)
Comment (puyuan1996) [reply lightrft/strategy/strategy_base.py:428]: 这个改动确实有效防止了 NCCL deadlock，但是，这里的实现有两个比较严重的隐患：

数学逻辑问题：如果 op="mean"，对于缺失 key 的 Rank 默认补 0.0，这会把 0.0 计入分子，并除以 world_size。这会严重拉低该指标的真实均值（比如只有一张卡有 KL=1.0，4卡平均后变成了 0.25）。
架构与性能问题：all_reduce 作为底层通信原语，内部高频调用 all_gather_object (依赖 pickle) 会带来性能损耗；而且在底层强行补 0.0 掩盖了上游状态不对齐的问题。

建议的修改方向：

最好不要在底层 all_reduce 中做 key 的对齐。我们应该在调用 all_reduce 之前的业务层（比如 metrics logging 处），显式地初始化所有可能的 keys。对于被 filter 掉的 rank，可以传 0.0 并配合一个 valid_count 掩码，最后用 sum(values) / sum(valid_counts) 来计算准确的 mean。

**Comment by Jiaxuan-Sun** [comment]
Need to be tested.
